lib/executor/executor.cpp

```diff
@@ -1,14 +1,14 @@
 #include "shell/executor/executor.hpp"
 
 #include "shell/logger.hpp"
-#include "shell/util.hpp"
+#include "shell/utility/types.hpp"
 #include <stdexcept>
 #include <cassert>
 #include <stdio.h>
 
 namespace shell {
 
-Executor::Executor(char *const *envp) : envp(envp) {
```

Comment:
I think it'll be nice to still pass the environment to classes through the constructor, to be able to test them easily  with custom environments

Reply:
I dont think this is a good idea, since the executor will be traversing the ast, and some commands may alter the env, it would be annoying to update the envp for the calls before.

Reply:
Passing it to the executor is fine, no? Then each node will simply use the same env that was passed through the constructor (i.e. the global one in the case of the actual program, but some local variant in the case of a test

Reply:
I think it should definitely be an Interface so we can mock it
I also think it should be passed to the Executor

Since we need the `envp` when executing a program, I think the Environment should be a shared pointer, whenever a modification is made to the environment it should make a new Environment, basically copy on write

Reply:
> The shell shall give the export attribute to the variables corresponding to the specified names, which shall cause them to be in the environment of subsequently executed commands. If the name of a variable is followed by = word, then the value of that variable shall be set to word.

https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_12

> Variables with the [export](https://pubs.opengroup.org/onlinepubs/009695399/utilities/export.html) attribute, along with those explicitly exported for the duration of the command, shall be passed to the utility environment variables

Reply:
This means that we could have multiple environments in one Execute call, for separate commands

When a command gets executed, we call getEnvironmentVariables, we likely want to cache this (later) in case commands get executed with no `export` calls inbetween or any transient variables are set (for example: `MY_KEY=MY_VALUE ./my_executable`)

Probably the `Environment` itself doesn't need to be copy-on-write, but the produced `envp` (from `getEnvironmentVariables`) should be a shared_ptr, the Environment keeps track of any modifications made since the last materialization - if no modifications are made it can return the same `envp`, if any modifications are made it will re-materialize the `envp` and hand that out instead.

Every command needs to hold a `shared_ptr<char*const*> envp` to make sure its version of the environment stays alive long enough

Reply:
Why copy on write? Aren't environment changes passed on to the next execution unit? Example:
```
export a=b ; echo $a # runs export in current shell environment
(export a=b) ;  echo $a # runs export in subshell, not exported to same environment
```
This behavior kind of seems like it should be a global, if it is exported or modified in a child process it won't be seen, which is correct.

Reply:
I didn't  see your next comment before I wrote mine, I see your point. I think it makes sense to support both in the end but I think the 'localised' environments are out of scope for this PR, making it an interface makes it easier to add later